module-4/backend/week_three.md

```diff
@@ -1,19 +1,52 @@
+
 3. How do we setup a route when creating an API with Node and Express? Please provide a code snippet.
+You could first make an Express instance:
+  `const express = require('express')`
+  `const app = express()`
+Then, call a method on the Express instance representing the HTTP verb of a request. Pass in arguments representing the path to match and the callback to invoke when that path is requested. For example:
+  `app.get('api/v1/foods', (request, response) => {
+      FoodController.all(response)
+    })`
+
 4. What do we use Knex for?
+Knex allows us to execute SQL code in a Node environment. This would be required for querying the database to create, read, update, and delete data.
+
 5. How could you organize your code to follow the MVC design pattern in your Quantified Self project?
+In Quantified Self, we could follow MVC by making a back-end API containing models and controllers. There would be a controllers folder which contains controllers, and the controller methods would be invoked within a callback in an Express route handler. The back-end app would also have a models folder which contains models that query the database in response to a controller action. The view would live in an entirely separate application that queries the back-end API via AJAX. This results in an architecture where the back-end application contains controllers and models, and the front-end application contains views.
+
 6. How do you execute raw SQL in node?
+You could first create a knex instance:
+  `const environment = process.env.NODE_ENV || 'development';`
+  `const configuration = require('../../knexfile')[environment];`
+  `const database = require('knex')(configuration);`
+
+Then, call .raw on that instance and pass in a string for the SQL query:
+  `database.raw('SELECT * FROM table_name')`
+
 7. What are some advantages to having your front end and back end code live in separate applications? What are some disadvantages?
+Advantages are better separation of concerns and better clarity about where to find the source code for different parts of your application. Disadvantages involve having to troubleshoot CORS issues, and possibly more time required to think about and architect the interface for the two applications.
 
 #### Review  
 
 8. Describe DNS.
+Domain Name Services resolves a URL to an IP address which represents a network and a host. This enables an HTTP request to get routed to the proper host.
+
 9. What does writing clean code mean to you?
+  * No repetition in code. Reuse of methods instead of writing repeated code.
+  * Single-responsibility entities and methods.
+  * Descriptive names for classes, methods, and variables.
+
 10. If you were building an application that models hotels, what classes would you need? What classes would be responsible for what functionality?
+Guest class: responsible for representing guests and storing their name, contact information, and billing information
```

Comment:
great start here 👍  i'd also have a hotel class that's composed of all the rooms